shared_model/backend/protobuf/impl/proposal.cpp

```diff
@@ -5,38 +5,67 @@
 
 #include "backend/protobuf/proposal.hpp"
 
+#include "backend/protobuf/transaction.hpp"
+
 namespace shared_model {
   namespace proto {
     using namespace interface::types;
 
-    Proposal::Proposal(Proposal &&o) noexcept
-        : NonCopyableProto(std::move(o.proto_)) {}
+    struct Proposal::Impl {
+      Impl(TransportType &&ref) : proto_(std::move(ref)) {}
+      Impl(const TransportType &ref) : proto_(ref) {}
+      TransportType proto_;
+      const std::vector<proto::Transaction> transactions_{[this] {
+        return std::vector<proto::Transaction>(
+            proto_.mutable_transactions()->begin(),
+            proto_.mutable_transactions()->end());
+      }()};
+
+      interface::types::BlobType blob_{[this] { return makeBlob(proto_); }()};
+
+      const interface::types::HashType hash_{
+          [this] { return crypto::DefaultHashProvider::makeHash(blob_); }()};
+    };
 
-    Proposal &Proposal::operator=(Proposal &&o) noexcept {
-      proto_ = std::move(o.proto_);
+    Proposal::Proposal(Proposal &&o) = default;
 
-      hash_.invalidate();
-      transactions_.invalidate();
-      blob_.invalidate();
+    Proposal::Proposal(const TransportType &ref) {
+      impl_ = std::make_unique<Proposal::Impl>(std::move(ref));
+    }
 
-      return *this;
+    Proposal::Proposal(TransportType &&ref) {
+      impl_ = std::make_unique<Proposal::Impl>(ref);
```

Comment:
std::move is needed here

Reply:
*std::forward

Reply:
TransportType is not a template parameter, so it is std::move